### PR TITLE
Added support for IMap.replaceAll() #14913

### PR DESCRIPTION
**Pull request by webashutosh**
Closes: https://github.com/hazelcast/hazelcast/issues/14913

### CODE REVIEW

**Comment by webashutosh** [comment]
@mmedenjak raised this one for adding implementation for `replaceAll()`. Could you please have a look? cc @Holmistr

**Comment by mmedenjak** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        function = in.readObject();
+    }
+
+    @Nullable
+    @Override
+    public EntryProcessor<K, V, V> getBackupProcessor() {
+        return null;
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java:70]: I think we should simply not override this method to execute the same on the backup replicas.

**Review by mmedenjak — COMMENTED** [review]
Looks good, one minor change needed.

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java (hunk 1 above)
Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java:70]: @mmedenjak - I guess I should do the same for `KeyValueConsumingEntryProcessor` (used in `forEach()` implementation) - can you please confirm?

https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java#L1123

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java:70]: I've checked the others. Since we don't set the new value, means we're not mutating the entry, which means it's read-only.
Now that I think about it, it may be the case that some users use `IMap.forEach` to mutate the value. We have two options then:
• we continue supporting only read-only use of `forEach`. This way, users mutating the values may get surprised but they can always switch to using regular EPs. Also, this way we don't introduce a massive overhead for read-only uses of `forEach`, meaning we don't create defensive copies, serialize after the lambda has been invoked, invoke entry listeners, WAN replication, etc. for every single value
• we defensively act as if the value was mutated. That means both read-only and mutating lambdas will work correctly.

Between the two, I'd go with option 1. It's a bit surprising but users can always simply switch to a regular EP to mutate the values and we avoid the prohibitively expensive overhead of defensively copying, serializing, notifying listeners, etc. for read-only uses of `forEach`.

In other words, you can leave that one as-is.

Comment (webashutosh) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java:70]: Agree with you. `forEach()` is a bit subjective one!

**Review by mmedenjak — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapEntryReplacingEntryProcessor.java]: I guess there's also the marker interface of `ReadOnly` which can be added to the lambda but... I'm not sure if everyone is aware of it and I'd rather keep this for another PR if we decide to change the current behaviour.

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
We also need a client side implementation of `replaceAll` in `ClientMapProxy` class. 
You can use [removeAll](https://github.com/hazelcast/hazelcast/blob/629f4774ba57e21707b677336743db6c61faee2e/hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java#L308) as an example. Otherwise, there will be inconsistent behaviour of `replaceAll` between client and server.

**Review by ahmetmircik — CHANGES_REQUESTED** [review]
We also need a client side implementation of `replaceAll` in `ClientMapProxy` class. 
You can use [removeAll](https://github.com/hazelcast/hazelcast/blob/629f4774ba57e21707b677336743db6c61faee2e/hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientMapProxy.java#L308) as an example. Otherwise, there will be inconsistent behaviour of `replaceAll` between client and server.

**Comment by mmedenjak** [comment]
@ahmetmircik unfortunately the client has no way of knowing the cluster version, so it cannot use the entry processor approach. It's better off using the default `replaceAll` implementation which will be slow, but correct.

Instead, we might add the codecs, tasks and other full-blown client support to send the task to a member, which can then act accordingly. Not sure if it's worth it for 4.1 or for another PR as an enhancement. WDYT?

**Comment by ahmetmircik** [comment]
@mmedenjak to implement replaceAll for all type of clients we need full-blown client support. If we only implement this for server side, there will be behaviour inconsistency(even on java client) as you mentioned. For public api users, seeing less surprise may be preferable. But maybe, at least documenting this inconsistency can be another good thing to do for now.

**Comment by mmedenjak** [comment]
True. I guess we started by fixing the default behaviour of methods like `computeIfAbsent` provided by JDK8 `j.u.Map` and `j.u.c.ConcurrentMap` (which simply looped forever by default) and continued by providing "more efficient" methods for other methods like `replaceAll`.

Although it would be nice to provide it for all clients, we were kinda forced into it by JDK providing inefficient or defective default implementations, not because users asked for it.

So, for now, I'm fine with documentation and as you proposed, the next step would be to introduce codecs, tasks, client protocol changes, everything, for better integration with clients. Maybe we can do that in 4.2 since we're now at the end of the release.

**Review by ahmetmircik — APPROVED** [review]

**Comment by mmedenjak** [comment]
Merged, thank you for the PR @webashutosh and review @ahmetmircik ! I'll open a new issue for the next steps.

**Comment by mmedenjak** [comment]
https://github.com/hazelcast/hazelcast/issues/17624
